### PR TITLE
DOC Add tutorial for adding a new package

### PR DESCRIPTION
**Pull request by rth**
This adds a more detailed tutorial on how to add new packages to Pyodide.

@sylvain-ri Does this answer some of your questions in https://github.com/pyodide/pyodide/issues/1724 ? If there is anything still unclear about that process please let me know.

[Rendered version](https://pyodide--1829.org.readthedocs.build/en/1829/development/new-packages.html)

### CODE REVIEW

File: docs/development/new-packages.md [hunk 1]
```diff
@@ -2,6 +2,101 @@
...
+PYODIDE_PACKAGES="<package-name>" make
+```
+
+and see if there are any errors. The detailed build log can be found under
+`packages/<package-name>/<package-name>.log`.
+
+If there are errors you might need to,
+
+- patch the package by adding `.patch` files to `packages/<package-name>/patches`
+- update the `meta.yaml`
```
Comment (hoodmane) [inline docs/development/new-packages.md:87]: Maybe more specifically 
```suggestion
- add the patch files to the `source/patches` field in the `meta.yaml` file
```

File: docs/development/meta-yaml.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+Packages are defined by writing a `meta.yaml` file. The format of these files is
+based on the `meta.yaml` files used to build [Conda
+packages](https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html),
+though it is much more limited. The most important limitation is that Pyodide
+assumes there will only be one version of a given library available, whereas
+Conda allows the user to specify the versions of each package that they want to
+install. Despite the limitations, keeping the file format as close as possible
+to conda's should make it easier to use existing conda package definitions as a
+starting point to create Pyodide packages. In general, however, one should not
+expect Conda packages to "just work" with Pyodide, see `pr`{795}`
```
Comment (hoodmane) [inline docs/development/meta-yaml.md:14]:
```suggestion
expect Conda packages to "just work" with Pyodide, see {pr}`795`
```

File: docs/development/meta-yaml.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+Packages are defined by writing a `meta.yaml` file. The format of these files is
+based on the `meta.yaml` files used to build [Conda
+packages](https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html),
+though it is much more limited. The most important limitation is that Pyodide
+assumes there will only be one version of a given library available, whereas
+Conda allows the user to specify the versions of each package that they want to
+install. Despite the limitations, keeping the file format as close as possible
+to conda's should make it easier to use existing conda package definitions as a
+starting point to create Pyodide packages. In general, however, one should not
```
Comment (hoodmane) [inline docs/development/meta-yaml.md:13]:
```suggestion
install. Despite the limitations, it is recommended to use existing conda 
package definitions as a starting point to create Pyodide packages. In
general, however, one should not
```

File: docs/development/meta-yaml.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+## `package`
+
+### `package/name`
+
+The name of the package. It must match the name of the package used when
+expanding the tarball, which is sometimes different from the name of the package
+in the Python namespace when installed. It must also match the name of the
+directory in which the `meta.yaml` file is placed. It can only contain
+alpha-numeric characters and `-`, `_`.
```
Comment (hoodmane) [inline docs/development/meta-yaml.md:26]:
```suggestion
alpha-numeric characters, `-`, and `_`.
```

File: docs/development/meta-yaml.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+### `build/cflags`
+
+Extra arguments to pass to the compiler when building for WebAssembly.
+
+(This key is not in the Conda spec).
+
+### `build/cxxflags`
+
+Extra arguments to pass to the compiler when building C++ files for WebAssembly. Note that both clfags and cxxflags will be used when compiling C++ files. A common example would be to use -std=c++11 for code that makes use of C++11 features.
```
Comment (hoodmane) [inline docs/development/meta-yaml.md:97]:
```suggestion
Extra arguments to pass to the compiler when building C++ files for WebAssembly.
Note that both `cflags` and `cxxflags` will be used when compiling C++ files. 
A common example would be to use `-std=c++11` for code that makes use of C++11 features.
```

File: docs/development/new-packages.md [hunk 6]
```diff
@@ -2,6 +2,101 @@
 
 # Creating a Pyodide package
 
+## Quickstart
+
+If you wish to use a package in Pyodide that is not already included, first you
+need to determine if it is necessary to package if for Pyodide. Ideally, you
```
Comment (hoodmane) [inline docs/development/new-packages.md:8]:
```suggestion
need to determine whether it is necessary to package it for Pyodide. Ideally, you
```

File: docs/development/new-packages.md [hunk 7]
```diff
@@ -2,6 +2,101 @@
...
+### 2. Creating the meta.yaml
+
+Once you determined that you need to create a new package for Pyodide, the
+easiest place to start is with the {ref}`mkpkg tool <pyodide-mkpkg>`. If your
+package is on PyPI, run:
+
+`pyodide-build mkpkg <package-name>`
+
+This will generate a `meta.yaml` under `package/<package-name>/` (see
+`ref`{meta-yaml-spec}) that should work out of the box for many simple Python
```
Comment (hoodmane) [inline docs/development/new-packages.md:52]:
```suggestion
{ref}`meta-yaml-spec`) that should work out of the box for many simple Python
```

File: docs/development/new-packages.md [hunk 8]
```diff
@@ -2,6 +2,101 @@
...
+`packages/<package-name>/<package-name>.log`.
+
+If there are errors you might need to,
+
+- patch the package by adding `.patch` files to `packages/<package-name>/patches`
+- update the `meta.yaml`
+
+then re-start the build.
+
+In general, it is useful to see how other similar packages are built in Pyodide.
```
Comment (hoodmane) [inline docs/development/new-packages.md:91]:
```suggestion
In general, it is recommended to look into how other similar packages are built in Pyodide.
```

**Review by hoodmane — APPROVED** [review]
Some minor edit ideas.

**Comment by sylvain-ri** [comment]
Hello. What is not clear for me is how to get the `pyodide build`. Is the hyphen missing (`pyodide-build`) ? Where do I get the pyodide build command ? It is mentionned on the [pyodide-build-cli page](https://pyodide--1829.org.readthedocs.build/en/1829/usage/api/pyodide-build-cli.html#pyodide-mkpkg) but not how to get it.

Would it easy to use the command for a wheel that is not on pypi as well ? If it's just an URL change/paremeter, that would be easier than uploading each time, especially for testing purposes (pypi enforces new version for each new upload).

**Comment by rth** [comment]
It's in the Pyodide repo. You can clone the repo, then run,
```bash
pip install -e pyodide-build
```
to install it. Also currently we use this CLI only for `mkpkg`, for actually building packages you still need to do,
```bash
PYODIDE_PACKAGES="<package-name>" make
```

> What is not clear for me is how to get the pyodide build. Is the hyphen missing (pyodide-build) ? 

Yes, it's "pyodide-build", where did you see it called "pyodide build"?

**Comment by sylvain-ri** [comment]
> This adds a more detailed tutorial on how to add new packages to Pyodide.
> 
> @sylvain-ri Does this answer some of your questions in #1724 ? If there is anything still unclear about that process please let me know.
> 
> [Rendered version](https://pyodide--1829.org.readthedocs.build/en/1829/development/new-packages.html)

Yes that's much better, thanks !

**Comment by sylvain-ri** [comment]
> > What is not clear for me is how to get the pyodide build. Is the hyphen missing (pyodide-build) ?
> 
> Yes, it's "pyodide-build", where did you see it called "pyodide build"?

I think this page confused me: 
https://pyodide.org/en/stable/usage/api/pyodide-build-cli.html
https://pyodide--1829.org.readthedocs.build/en/1829/usage/api/pyodide-build-cli.html#pyodide-mkpkg

**Comment by rth** [comment]
Indeed, we should probably replace "Make a new pyodide package." there by "Generate a `meta.yaml`  for a new Pyodide package"

**Comment by sylvain-ri** [comment]
> It's in the Pyodide repo. You can clone the repo, then run,
> 
> ```shell
> pip install -e pyodide-build
> ```

Ooh that's would definitely help to have is directly in the docs ! Let me try to add it.

> to install it. Also currently we use this CLI only for `mkpkg`, for actually building packages you still need to do,
> 
> ```shell
> PYODIDE_PACKAGES="<package-name>" make
> ```

The make command failed for me because of dependencies. Should we add a `pip install -r requirements.txt` in the make file ?

**Comment by sylvain-ri** [comment]
> Indeed, we should probably replace "Make a new pyodide package." there by "Generate a `meta.yaml` for a new Pyodide package"

That will help. So this is to add a package to pyodide natively supported packages ? Those : https://github.com/pyodide/pyodide/tree/main/packages ?

**Comment by rth** [comment]
> Ooh that's would definitely help to have is directly in the docs ! Let me try to add it.

Thanks! Generally don't hesitate to open PRs to improve anything you find unclear in the docs. In your commit messages it's useful to append ` [skip ci]` to skip build CI which takes a while and is not relevant for documentation updates.

> The make command failed for me because of dependencies. Should we add a pip install -r requirements.txt in the make file ?

Yes, we should improve https://pyodide.org/en/latest/development/building-from-sources.html as well

**Comment by rth** [comment]
> So this is to add a package to pyodide natively supported packages ?

Yes, it will create the `packages/<name>/meta.yaml` file, that's it. No guarantee that this recipe will compile without changes though, particularly for more complex packages. So then one need to build it with `make` to be sure.

**Comment by sylvain-ri** [comment]
> Thanks! Generally don't hesitate to open PRs to improve anything you find unclear in the docs. In your commit messages it's useful to append ` [skip ci]` to skip build CI which takes a while and is not relevant for documentation updates.

Ok thanks. First time contributing to another project, so let me know if i do any mistake.

**Comment by sylvain-ri** [comment]
Sorry for the stupid question: i should commit on a new branch or directly on the master branch ? I couldn't find a way to create a new branch in github, or to push the branch i created locally.

**Comment by rth** [comment]
Creating a new branch for each PR is recommended. The usual steps are below,
```
git checkout main  # to make sure you create it from your main branch
git checkout -b my-new-branch
git push
```

**Comment by hoodmane** [comment]
I think if you just say
```
git checkout -b new-branch
git push
```
it will say:
```
fatal: The current branch new-branch has no upstream branch.
```
Is there some git setting that causes it to automatically make a new tracking upstream branch? I usually would say `git push -u origin new-branch`.

**Comment by hoodmane** [comment]
Also, @sylvain-ri, I think you need to set up a fork first. I think this guide might be helpful:
https://www.freecodecamp.org/news/how-to-make-your-first-pull-request-on-github-3/

Is there a standard reference for this? We might want to link a tutorial like this in https://pyodide.org/en/stable/development/contributing.html

**Pull request merged**